### PR TITLE
reauth

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Q Reauth

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/16214713-6e18-4ce9-bddf-c5e98bf74b33

## Q Logout

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/47e9f581-af83-44e8-a6ff-ffa39a4cab6f

## Q Functional test

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/d9b3d3ea-6bb1-4ecc-a204-18157af4c14f

## Will not show if there is IAM credential (show if and only if there is only codecatalyst connection and it's expired)

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/55ae9ad0-c74a-42eb-ac7f-1c2d8b209058

## Backbutton to go back to explorer tree

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/c6ea0b38-d68f-4f4e-b045-de58dfb60d1d

## Signout button

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/adc21886-e372-48b6-8580-a811a1313ae2

## reauth button

https://github.com/aws/aws-toolkit-jetbrains/assets/96078566/c9ca8d6d-046c-4c68-b8d1-5d1122ca53a9

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindow.kt [hunk 1]
```diff
@@ -32,17 +27,6 @@ import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.auth.isFeature
 import software.aws.toolkits.jetbrains.services.codemodernizer.utils.isCodeTransformAvailable
 import javax.swing.JComponent
 
-// TODO: move to somewhere else utils
-fun isQConnected(project: Project): Boolean {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindow.kt:36]: moved

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt [hunk 2]
```diff
@@ -155,7 +157,13 @@ class QWebviewBrowser(val project: Project, private val parentDisposable: Dispos
             }
 
             "reauth" -> {
-                // TODO: implementation
+                ToolkitConnectionManager.getInstance(project).activeConnectionForFeature(QConnection.getInstance())?.let { conn ->
+                    if (conn is ManagedBearerSsoConnection) {
+                        ApplicationManager.getApplication().executeOnPooledThread {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/QLoginWebview.kt:162]: forgot to use `loginWithBackgroundContext`

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt [hunk 3]
```diff
@@ -47,8 +43,7 @@ abstract class LoginBrowser(
     protected data class BearerConnectionSelectionSettings(val currentSelection: AwsBearerTokenConnection, val onChange: (AwsBearerTokenConnection) -> Unit)
     protected val selectionSettings = mutableMapOf<String, BearerConnectionSelectionSettings>()
 
-    // TODO: figure out a better way to do this UI update
-    protected val onPendingProfile: (InteractiveBearerTokenProvider) -> Unit = { provider ->
```
Comment (Will-ShaoHua) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt:51]: renamed

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt [hunk 4]
```diff
@@ -58,25 +53,6 @@ abstract class LoginBrowser(
...
-                val authorization = pollForAuthorization(provider as InteractiveBearerTokenProvider)
-
-                if (authorization != null) {
-                    executeJS("window.ideClient.updateAuthorization(\"${userCodeFromAuthorization(authorization)}\")")
-                    currentAuthorization = authorization
-                    return@launch
-                }
-            }
-        }
-    }
```
Comment (Will-ShaoHua) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt:78]: removed

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt [hunk 5]
```diff
@@ -100,13 +76,9 @@ abstract class LoginBrowser(
         else -> ""
     }
 
-    fun resetBrowserState() {
-        executeJS("window.ideClient.reset()")
-    }
```
Comment (Will-ShaoHua) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt:105]: not being used, add back when needed

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt [hunk 6]
```diff
@@ -178,10 +158,6 @@ abstract class LoginBrowser(
         """.trimIndent()
     }
 
-    protected suspend fun pollForConnection(connectionId: String): ToolkitConnection? = pollFor {
-        ToolkitAuthManager.getInstance().getConnection(connectionId)
-    }
-
```
Comment (Will-ShaoHua) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/webview/LoginBrowser.kt:184]: not being used, removed

File: plugins/core/webview/src/q-ui/components/login.vue [hunk 7]
```diff
@@ -87,50 +65,20 @@ export default defineComponent({
...
-                return 'Connecting to IAM...'
-            } else if (this.selectedLoginOption?.id === LoginIdentifier.BUILDER_ID || this.selectedLoginOption?.id === LoginIdentifier.ENTERPRISE_SSO) {
-                return 'Authenticating in browser...'
-            }
-
-            return ''
-        },
-        requireConfirmationCodeOrNot(): boolean {
-            return this.selectedLoginOption?.requiresBrowser() === true && this.authorizationCode?.length !== 0
-        }
```
Comment (Will-ShaoHua) [inline plugins/core/webview/src/q-ui/components/login.vue:112]: moved to authenticating.vue

File: plugins/core/webview/src/q-ui/components/login.vue [hunk 8]
```diff
@@ -87,50 +65,20 @@ export default defineComponent({
...
         handleCancelButton() {
             window.ideClient.cancelLogin()
             this.mutateStage('START')
-            this.authorizationCode = undefined
-        },
-        changeTheme(darkMode: boolean) {
-            const oldCssId = darkMode ? "jb-light" : "jb-dark"
-            const newCssId = darkMode ? "jb-dark" : "jb-light"
-            document.body.classList.add(newCssId);
-            document.body.classList.remove(oldCssId);
```
Comment (Will-ShaoHua) [inline plugins/core/webview/src/q-ui/components/login.vue:133]: moved to root.vue

Comment (Will-ShaoHua) [reply plugins/core/webview/src/q-ui/components/login.vue:133]: https://github.com/aws/aws-toolkit-jetbrains/pull/4370/files#diff-d831ab0cf73aa597f4bf0c976d638e38dc9abb52042ceea0bc6b6b81439b0bc9R43-R54

File: plugins/core/webview/src/q-ui/components/login.vue [hunk 9]
```diff
@@ -154,36 +102,14 @@ export default defineComponent({
             } else if (type instanceof ExistConnection) {
                 window.ideApi.postMessage({ command: 'selectConnection', connectionId:  type.pluginConnectionId})
             }
-        }
-    },
-    mounted() {
-        console.log('login mounted')
-        window.changeTheme = this.changeTheme.bind(this)
-        window.ideApi.postMessage({command: 'prepareUi'})
```
Comment (Will-ShaoHua) [inline plugins/core/webview/src/q-ui/components/login.vue:162]: moved to root.vue
https://github.com/aws/aws-toolkit-jetbrains/pull/4370/files#diff-d831ab0cf73aa597f4bf0c976d638e38dc9abb52042ceea0bc6b6b81439b0bc9R43-R54

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt [hunk 10]
```diff
@@ -39,7 +43,28 @@ class AmazonQToolWindowFactory : ToolWindowFactory, DumbAware {
...
 
-        val component = if (isQConnected(project)) {
+        ApplicationManager.getApplication().messageBus.connect().subscribe(
+            BearerTokenProviderListener.TOPIC,
+            object : BearerTokenProviderListener {
+                override fun onChange(providerId: String, newScopes: List<String>?) {
+                    println("providerId: $providerId")
+                    if ((isQConnected(project) && !isQExpired(project))) {
+                        println("providerId: $providerId connected")
+                        val content = contentManager.factory.createContent(AmazonQToolWindow.getInstance(project).component, null, false).also {
```
[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt:53]: won't this cause any memory leak beacuse project got disposed but application is not?

File: plugins/core/webview/src/q-ui/components/reauth.vue [hunk 11]
```diff
@@ -2,57 +2,92 @@
...
-        v-on:click="handleContinueClick()"
-    >
-        Re-authenticate
+    <button v-if="!this.authenticating && this.app === 'TOOLKIT'" class="back-button" @click="back" tabindex="-1">
+        <svg width="24" height="24" viewBox="0 -3 13 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <path
+                d="M4.98667 0.0933332L5.73333 0.786666L1.57333 4.94667H12.0267V5.96H1.57333L5.73333 10.0667L4.98667 10.8133L0.0266666 5.8V5.10667L4.98667 0.0933332Z"
+                fill="#21A2FF"
+            />
+        </svg>
```
Comment (andrewyuq) [inline plugins/core/webview/src/q-ui/components/reauth.vue:11]: Will this again generate the same issue that has two back buttons?

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt [hunk 12]
```diff
@@ -100,7 +105,20 @@ class AwsToolkitExplorerFactory : ToolWindowFactory, DumbAware {
...
+
+        ApplicationManager.getApplication().messageBus.connect().subscribe(
+            BearerTokenProviderListener.TOPIC,
+            object : BearerTokenProviderListener {
+                override fun onChange(providerId: String, newScopes: List<String>?) {
+                    if (ToolkitConnectionManager.getInstance(project)
+                            .connectionStateForFeature(CodeCatalystConnection.getInstance()) == BearerTokenAuthState.AUTHORIZED
+                    ) {
+                        showExplorerTree(project)
+                    }
```
Comment (andrewyuq) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:121]: same, stale project should not live beyond project scope

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt:53]: umm, should we add `parentDisposable=project`?

File: plugins/core/webview/src/q-ui/components/reauth.vue (hunk 11 above)
Comment (Will-ShaoHua) [reply plugins/core/webview/src/q-ui/components/reauth.vue:11]: nope, `reauth.vue` now is independent to `login.vue`

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt (hunk 12 above)
Comment (Will-ShaoHua) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:121]: add `parentDisposable=project`?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt:53]: how do we ensure Richard's comment `if we need to protect this component from being disposed`

File: plugins/core/webview/src/q-ui/components/reauth.vue (hunk 11 above)
Comment (Will-ShaoHua) [reply plugins/core/webview/src/q-ui/components/reauth.vue:11]: so in `root.vue`
```
<template>
    <Login v-if>
    <Reauth v-else>
</template>
```

Comment (Will-ShaoHua) [reply plugins/core/webview/src/q-ui/components/reauth.vue:11]: backbutton lives within Login and Reauth respectively

Comment (andrewyuq) [reply plugins/core/webview/src/q-ui/components/reauth.vue:11]: got it

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrewyuq) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt:53]: We just need to make sure every createContent needs to have a removeAllContent(true) paired somewhere and it needs to be guranteed to happen.
For the memory leak, just do project.messageBus

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt (hunk 12 above)
Comment (andrewyuq) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:121]: project should not be chosen as a parent disposable
https://plugins.jetbrains.com/docs/intellij/disposers.html#choosing-a-disposable-parent

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/toolwindow/AmazonQToolWindowFactory.kt:53]: sg, was not aware of project.messagebus can subscribe to app level topic

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt (hunk 12 above)
Comment (Will-ShaoHua) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:121]: gotcha

Comment (Will-ShaoHua) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:121]: didn't get a chance to look into disposable yet

**Review by andrewyuq — APPROVED** [review]

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt [hunk 13]
```diff
@@ -100,7 +104,20 @@ class AwsToolkitExplorerFactory : ToolWindowFactory, DumbAware {
...
+        )
+
+        project.messageBus.connect().subscribe(
+            BearerTokenProviderListener.TOPIC,
+            object : BearerTokenProviderListener {
+                override fun onChange(providerId: String, newScopes: List<String>?) {
+                    if (ToolkitConnectionManager.getInstance(project)
+                            .connectionStateForFeature(CodeCatalystConnection.getInstance()) == BearerTokenAuthState.AUTHORIZED
+                    ) {
+                        showExplorerTree(project)
```
Comment (manodnyab) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/AwsToolkitExplorerFactory.kt:119]: I wonder if we can make this default to the dev tools tab

**Review by manodnyab — APPROVED** [review]

**Pull request merged**